packages/amplify_core/lib/src/types/models/mipr/schema_type.dart

```diff
@@ -0,0 +1,644 @@
+  static AppSyncScalar valueOf(String name) => _$AppSyncScalarValueOf(name);
+
+  /// Bridge for legacy codegen.
+  // TODO(dnys1): Remove with platform channel code
+  static AppSyncScalar? fromLegacyType(ModelFieldTypeEnum type) {
+    switch (type) {
+      case ModelFieldTypeEnum.string:
+        return string;
+      case ModelFieldTypeEnum.int:
+        return int_;
+      case ModelFieldTypeEnum.double:
+        return float;
+      case ModelFieldTypeEnum.date:
+        return awsDate;
+      case ModelFieldTypeEnum.dateTime:
+        return awsDateTime;
+      case ModelFieldTypeEnum.time:
+        return awsTime;
+      case ModelFieldTypeEnum.timestamp:
+        return awsTimestamp;
+      case ModelFieldTypeEnum.bool:
+        return boolean;
+      case ModelFieldTypeEnum.enumeration:
+      case ModelFieldTypeEnum.model:
+      case ModelFieldTypeEnum.collection:
+      case ModelFieldTypeEnum.embedded:
+      case ModelFieldTypeEnum.embeddedCollection:
+        return null;
+    }
+  }
+
+  /// The serializer for [AppSyncScalar].
+  static Serializer<AppSyncScalar> get serializer => _$appSyncScalarSerializer;
+}
+
+/// {@template amplify_core.models.schema_type}
+/// An Amplify schema type, either a scalar, model, non-model, or enum.
+///
+/// - [ScalarType]
+/// - [ModelType]
+/// - [NonModelType]
+/// - [EnumType]
+/// - [ListType]
+/// {@endtemplate}
+abstract class SchemaType with AWSSerializable<Map<String, Object?>> {
+  /// {@macro amplify_core.models.schema_type}
+  const SchemaType();
+
+  /// {@template amplify_core.models.scalar_type}
+  /// A scalar (primitive) type, e.g. `ID`, `String`, `AWSDateTime`.
+  ///
+  /// AppSync defines 9 [scalar types](https://docs.aws.amazon.com/appsync/latest/devguide/scalars.html)
+  /// on top of the default 5 GraphQL scalar types (`ID`, `String`, `Int`,
+  /// `Float`, `Boolean`):
+  /// - [AppSyncScalar.awsDate]
+  /// - [AppSyncScalar.awsTime]
+  /// - [AppSyncScalar.awsDateTime]
+  /// - [AppSyncScalar.awsTimestamp]
+  /// - [AppSyncScalar.awsEmail]
+  /// - [AppSyncScalar.awsJson]
+  /// - [AppSyncScalar.awsPhone]
+  /// - [AppSyncScalar.awsUrl]
+  /// - [AppSyncScalar.awsIpAddress]
+  /// {@endtemplate}
+  const factory SchemaType.scalar(
+    AppSyncScalar value, {
+    bool isRequired,
+  }) = ScalarType._;
+
+  /// {@template amplify_core.models.model_type}
+  /// A custom structure with an `@model` directive.
+  ///
+  /// ```graphql
+  /// # `MyModel` is a model type.
+  /// type MyModel @model {
+  ///   id: ID!
+  /// }
+  /// ```
+  /// {@endtemplate}
+  const factory SchemaType.model(
+    String name, {
+    bool isRequired,
+  }) = ModelType._;
+
+  /// {@template amplify_core.models.non_model_type}
+  /// A custom structure with no `@model` directive.
+  ///
+  /// ```graphql
+  /// # `MyNonModel` is a non-model type.
+  /// type MyNonModel {
+  ///   name: String!
+  /// }
+  /// ```
+  /// {@endtemplate}
+  const factory SchemaType.nonModel(
```

Comment:
I couldn't find non-model references from both legacy and transformer v2 backed data modeling documentation...
So I'd recommend to name this type `customType` following the name used by Amplify Studio which is a public facing interface. 

<img width="612" alt="image" src="https://user-images.githubusercontent.com/10602282/192661388-56f80f01-c826-461e-9d75-7a67892b990a.png">

Reply:
NonModel is how they call it in amplify-codegen. Also, I think customType is kind of misleading - all model, non-model, and enum types are custom types.

Reply:
For public APIs we can keep it customType, but this would be an internal API.

Reply:
Non-model could be type other than model including scalar, enum ha! confusing either way. I agree we only need to keep the naming consistent on public APIs. (as long as it's not `embedded` 😅)

Reply:
Haha true, it's not much better. Only in the sense that it would align with codegen at this layer.